src/main/java/frc/robot/subsystems/swerve/states/SwerveState.java

```diff
@@ -0,0 +1,119 @@
+package frc.robot.subsystems.swerve.states;
+
+import frc.robot.subsystems.swerve.states.aimassist.AimAssist;
+import frc.robot.subsystems.swerve.states.heading.HeadingControl;
+
+public class SwerveState {
+
+	private static final DriveRelative DEFAULT_DRIVE_RELATIVE = DriveRelative.FIELD_RELATIVE;
+	private static final DriveSpeed DEFAULT_DRIVE_SPEED = DriveSpeed.NORMAL;
+	private static final LoopMode DEFAULT_LOOP_MODE = LoopMode.CLOSED;
+	private static final RotateAxis DEFAULT_ROTATE_AXIS = RotateAxis.MIDDLE_OF_CHASSIS;
+	private static final AimAssist DEFAULT_AIM_ASSIST = AimAssist.NONE;
+	private static final HeadingControl DEFAULT_HEADING_CONTROL = HeadingControl.NONE;
+
+	public static final SwerveState DEFAULT_PATH_PLANNER = new SwerveState().withDriveRelative(DriveRelative.ROBOT_RELATIVE);
+	public static final SwerveState DEFAULT_DRIVE = new SwerveState();
+
+
+	private DriveRelative driveRelative;
+	private DriveSpeed driveSpeed;
+	private LoopMode loopMode;
+	private RotateAxis rotateAxis;
+	private AimAssist aimAssist;
+	private HeadingControl headingControl;
+
+	public SwerveState(SwerveState swerveState) {
+		this(
+			swerveState.driveRelative,
+			swerveState.driveSpeed,
+			swerveState.loopMode,
+			swerveState.rotateAxis,
+			swerveState.aimAssist,
+			swerveState.headingControl
+		);
+	}
+
+	private SwerveState() {
+		this(DEFAULT_DRIVE_RELATIVE, DEFAULT_DRIVE_SPEED, DEFAULT_LOOP_MODE, DEFAULT_ROTATE_AXIS, DEFAULT_AIM_ASSIST, DEFAULT_HEADING_CONTROL);
+	}
+
+	private SwerveState(
+		DriveRelative driveRelative,
+		DriveSpeed driveSpeed,
+		LoopMode loopMode,
+		RotateAxis rotateAxis,
+		AimAssist aimAssist,
+		HeadingControl headingControl
+	) {
+		this.driveRelative = driveRelative;
+		this.driveSpeed = driveSpeed;
+		this.loopMode = loopMode;
+		this.rotateAxis = rotateAxis;
+		this.aimAssist = aimAssist;
+		this.headingControl = headingControl;
+	}
+
+
+	public SwerveState withDriveRelative(DriveRelative driveRelative) {
```

Comment:
when we do a "with" constructor in our code it changes the same instance and doesn't create another one. it's better because it saves memory space

Reply:
because of the usage of DEFAULT_DRIVE and DEFAULT_PATH_PLANNER u can't use the same instance. U need same starting point when using the state.
EX:
getDriveByState(DEFAULT_DRIVE.withAimAssist(NOTE));
getDriveByState(DEFAULT_DRIVE.withLoopMode(OPEN));
i dont want the first command to affect the second